File: projecteuler.net/004-largest-palindrome-product.js
// http://jsfiddle.net/t3NHM/

// PROBLEM //

/*
A palindromic number reads the same both ways. The largest palindrome
made from the product of two 2-digit numbers is 9009 = 91 * 99.

Find the largest palindrome made from the product of two 3-digit numbers.
*/

// HELPERS //

Array.range=function(s,e){var a=[];while(s<e)a.push(s++);return a}
String.prototype.reverse=function(){return this.split('').reverse().join('')}
String.prototype.isPalindrome=function(){return this==this.reverse()}

// ANSWER //

var findLargestPalindromeProduct = function(start, end) {
    return Array.range(start, end).reduce(function(largest, number1) {
        return Array.range(number1, end).reduce(function(largest, number2) {
            var product = number1 * number2;
            if (product.toString().isPalindrome() && product > largest)
                largest = product;
            return largest;
        }, largest);
    }, 0);
}

var result = findLargestPalindromeProduct(100, 1000);

// 906609

File: projecteuler.net/004-largest-palindrome-product.fs
// PROBLEM //

(*
A palindromic number reads the same both ways. The largest palindrome
made from the product of two 2-digit numbers is 9009 = 91 * 99.

Find the largest palindrome made from the product of two 3-digit numbers.
*)

// ANSWER //
module Program

open System.Linq

let isPalindrome n =
  let charArray = n.ToString().ToCharArray()
  let reverseCharArray = Array.rev charArray
  charArray.SequenceEqual(reverseCharArray)

let numbers = [100..999]
let products =
  numbers
  |> List.collect (fun x -> numbers |> List.map(fun y -> x * y))

let result =
  products
  |> Seq.filter isPalindrome
  |> Seq.max

// val result : int = 906609
